## Data Cleaning
------------------
<div align='justify'>For this part, we will try to understand West Texas Intermediate (WTI) price data that we crawled from <a href="https://markets.businessinsider.com/">Business Insider</a> webiste. Null values will be detected and removed if the frequency of the null values is small.

Let us start with reading the crawled data from my <a href="https://github.com/muhdamirulsamsul/DataMining">GitHub repository</a> using `pandas.read_csv()`. But before that, we will have to import all the necessary libraries.</div>

In [1]:
import pandas as pd


Read the csv file from the following link:
`https://raw.githubusercontent.com/muhdamirulsamsul/DataMining/master/wtiprice.csv`
<br/>and assign it to variable named `df`.

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/muhdamirulsamsul/DataMining/master/wtiprice.csv')

Let us have a glimpse of how WTI price data looks. Simply use `DataFrame.head()` to return the first 5 rows of the dataframe and `DataFrame.tail()` to return the last 5 rows. 

In [3]:
df.head()

Date  Closing Price   Open  Daily High  Daily Low
0  2020-03-11          32.98  34.62       36.35      32.56
1  2020-03-10          34.36  30.37       35.02      30.20
2  2020-03-09          31.13  32.87       34.88      27.34
3  2020-03-06          41.28  46.09       46.38      41.05
4  2020-03-05          45.90  47.13       47.57      45.65

In [4]:
df.tail()

Date  Closing Price   Open  Daily High  Daily Low
3609  2006-02-09          62.62  62.68       63.73      62.38
3610  2006-02-08          62.55  62.96       63.44      62.29
3611  2006-02-07          63.09  64.82       64.90      62.81
3612  2006-02-06          65.11  66.35       66.50      64.77
3613  2006-02-03          65.37  64.77       65.48      63.93

By looking at the first 5 and last 5 rows, we now have a little idea how the data looks like. Basically the data have 5 columns which are:

|Column|Description|
|------|-----------|
|Date| The date on which the data point is recorded|
|Closing Price| WTI price at the end of the day's business in a financial market| 
|Open| The price of the first trade for WTI|
|Daily High| The highest WTI price during the coruse of the trading day|
|Daily Low| The lowest WTI price during the course of the trading day|

We can display the information about the dataframe using the `DataFrame.info()` method.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3614 entries, 0 to 3613
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           3614 non-null   object 
 1   Closing Price  3614 non-null   float64
 2   Open           3554 non-null   float64
 3   Daily High     3595 non-null   float64
 4   Daily Low      3602 non-null   float64
dtypes: float64(4), object(1)
memory usage: 141.3+ KB


From the information displayed above, we can see that the dataframe has `3614` entries, and the index of the row is ranging from `0` to `3613`. However, three columns from the dataframe have missing values which are:
* `Open`
* `Daily High`
* `Daily Low`

Before we handle the missing values, it is easier for us if the column labels is consistent. We will clean our column labels by:
* Replacing spaces with underscores.
* Making all the characters lowercase.
* Shortening any long column names.

Below is the function to make the column labels easier to work with:

In [6]:
def clean(a):
    c = []
    for b in a:
        b = b.lower()
        b = b.replace(' ', '_')
        if b == 'open':
            b = b+'ing_price'
        c.append(b)
    return c

In [7]:
df.columns = clean(df.columns)
df.columns

Index(['date', 'closing_price', 'opening_price', 'daily_high', 'daily_low'], dtype='object')

Previously, some columns are identified to have missing values. These missing values are called null values. In pandas, null values are indicated by either NaN or None. We can use `DataFrame.isnull()` method to identify missing values, which returns a boolean dataframe. Then, we can use the `DataFrame.sum()` method to give the count of `True` values for each column: 

In [10]:
df.isnull()
# this will return a boolean dataframe

date  closing_price  opening_price  daily_high  daily_low
0     False          False          False       False      False
1     False          False          False       False      False
2     False          False          False       False      False
3     False          False          False       False      False
4     False          False          False       False      False
...     ...            ...            ...         ...        ...
3609  False          False          False       False      False
3610  False          False          False       False      False
3611  False          False          False       False      False
3612  False          False          False       False      False
3613  False          False          False       False      False

[3614 rows x 5 columns]

In [9]:
df.isnull().sum()

date              0
closing_price     0
opening_price    60
daily_high       19
daily_low        12
dtype: int64

We can see that `opening_price` has `60` null values, `daily_high` has `19` null values, and `daily_low` has `12` null values. Let us have a glimpse on which rows that null values resides.

In [15]:
df.loc[(df.opening_price.isnull())|(df.daily_high.isnull())|(df.daily_low.isnull()),:]

date  closing_price  opening_price  daily_high  daily_low
13    2020-02-21          53.38            NaN         NaN      52.55
16    2020-02-18          52.05            NaN         NaN      50.88
34    2020-01-22          56.74            NaN         NaN      56.03
54    2019-12-20          60.44            NaN       61.20      60.02
74    2019-11-21          58.58            NaN       58.67      56.60
...          ...            ...            ...         ...        ...
3045  2008-04-22         118.07            NaN      118.99     115.95
3068  2008-03-19         102.54            NaN      107.79     101.81
3088  2008-02-20          99.70            NaN      100.72      97.96
3131  2007-12-19          91.24            NaN       92.32      89.70
3153  2007-11-16          93.96            NaN       94.03        NaN

[63 rows x 5 columns]

The number of rows that have missing values are only 63, thus deleting the 63 rows can be considered. We can drop all the null values from the dataframe using `DataFrame.dropna()` method as follows:

In [19]:
df = df.dropna()

We can see the number of rows is changed from `3614` to `3551`. Try running df.isnull().sum() one more time just to make sure that there is no more null values.

In [21]:
df.isnull().sum()

date             0
closing_price    0
opening_price    0
daily_high       0
daily_low        0
dtype: int64